File: discord-bot-fleet/src/llm.ts
// OpenAI-compatible LLM client. Works with your proxy (lolmaobruhhh-fap.hf.space/v1, FAP!).
// Supports tool calling via the standard OpenAI function-calling protocol.

import OpenAI from 'openai';
import type { BotConfig } from './types.js';
import { decryptString } from './crypto.js';

export interface LLMMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string;
  name?: string;
  tool_call_id?: string;
  tool_calls?: Array<{
    id: string;
    type: 'function';
    function: { name: string; arguments: string };
  }>;
}

export interface LLMTool {
  type: 'function';
  function: {
    name: string;
    description: string;
    parameters: Record<string, unknown>; // JSON schema
  };
}

export interface LLMResponse {
  content: string | null;
  tool_calls?: LLMMessage['tool_calls'];
  finish_reason: string;
  usage?: { prompt_tokens: number; completion_tokens: number; total_tokens: number };
}

export class LLMClient {
  private client: OpenAI;
  private model: string;
  private temperature: number;
  private maxTokens: number;

  constructor(config: BotConfig) {
    // Decrypt API key. If empty, fall back to LLM_API_KEY env var (HF Secret).
    let apiKey: string;
    if (config.llm.api_key_enc) {
      apiKey = decryptString(config.llm.api_key_enc);
    } else {
      apiKey = process.env.LLM_API_KEY || 'FAP!';
    }

    this.client = new OpenAI({
      baseURL: config.llm.proxy_url,
      apiKey,
    });
    this.model = config.llm.model;
    this.temperature = config.llm.temperature;
    this.maxTokens = config.llm.max_tokens;
  }

  async chat(
    messages: LLMMessage[],
    tools: LLMTool[] = [],
    signal?: AbortSignal
  ): Promise<LLMResponse> {
    const params: Record<string, unknown> = {
      model: this.model,
      messages: messages as OpenAI.Chat.Completions.ChatCompletionMessageParam[],
      temperature: this.temperature,
      max_tokens: this.maxTokens,
    };
    if (tools.length > 0) {
      params.tools = tools as OpenAI.Chat.Completions.ChatCompletionTool[];
    }
    if (signal) params.signal = signal;

    const resp = await this.client.chat.completions.create(params as unknown as OpenAI.Chat.Completions.ChatCompletionCreateParamsNonStreaming);
    const choice = resp.choices[0];
    return {
      content: choice.message.content,
      tool_calls: choice.message.tool_calls as LLMMessage['tool_calls'] | undefined,
      finish_reason: choice.finish_reason,
      usage: resp.usage ? {
        prompt_tokens: resp.usage.prompt_tokens,
        completion_tokens: resp.usage.completion_tokens,
        total_tokens: resp.usage.total_tokens,
      } : undefined,
    };
  }

  // Test the proxy: simple "hello" request, returns latency and model list
  static async ping(proxyUrl: string, apiKey: string, model?: string): Promise<{
    status: 'online' | 'offline' | 'error';
    latency_ms?: number;
    models?: string[];
    test_response?: string;
    error?: string;
  }> {
    const start = Date.now();
    try {
      const client = new OpenAI({ baseURL: proxyUrl, apiKey });
      const modelsResp = await client.models.list();
      const models = Array.isArray(modelsResp.data) ? modelsResp.data.map(m => m.id) : [];
      let test_response: string | undefined;
      if (model) {
        const chat = await client.chat.completions.create({
          model,
          messages: [{ role: 'user', content: 'Say "hello" in one word.' }],
          max_tokens: 10,
        });
        test_response = chat.choices[0]?.message?.content || '';
      }
      return {
        status: 'online',
        latency_ms: Date.now() - start,
        models,
        test_response,
      };
    } catch (e: unknown) {
      const err = e as Error;
      return {
        status: 'error',
        latency_ms: Date.now() - start,
        error: err.message || String(e),
      };
    }
  }
}

File: discord-bot-fleet/bot-python/llm.py
"""OpenAI-compatible async LLM client with tool calling."""

import os
from typing import Any, Dict, List, Optional

from openai import AsyncOpenAI

from crypto import decrypt_string


class LLMClient:
    def __init__(self, config: Dict[str, Any]):
        if config['llm'].get('api_key_enc'):
            api_key = decrypt_string(config['llm']['api_key_enc'])
        else:
            api_key = os.environ.get('LLM_API_KEY', 'FAP!')
        self.client = AsyncOpenAI(
            base_url=config['llm']['proxy_url'],
            api_key=api_key,
            timeout=120.0,
        )
        self.model = config['llm']['model']
        self.temperature = config['llm']['temperature']
        self.max_tokens = config['llm']['max_tokens']

    async def chat(
        self,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Dict[str, Any]]] = None,
    ) -> Dict[str, Any]:
        kwargs: Dict[str, Any] = {
            'model': self.model,
            'messages': messages,
            'temperature': self.temperature,
            'max_tokens': self.max_tokens,
        }
        if tools:
            kwargs['tools'] = tools
        resp = await self.client.chat.completions.create(**kwargs)
        choice = resp.choices[0]
        tool_calls = None
        if choice.message.tool_calls:
            tool_calls = [
                {
                    'id': tc.id,
                    'type': 'function',
                    'function': {
                        'name': tc.function.name,
                        'arguments': tc.function.arguments,
                    },
                }
                for tc in choice.message.tool_calls
            ]
        usage = None
        if resp.usage:
            usage = {
                'prompt_tokens': resp.usage.prompt_tokens,
                'completion_tokens': resp.usage.completion_tokens,
                'total_tokens': resp.usage.total_tokens,
            }
        return {
            'content': choice.message.content,
            'tool_calls': tool_calls,
            'finish_reason': choice.finish_reason,
            'usage': usage,
        }

    @staticmethod
    async def ping(proxy_url: str, api_key: str, model: Optional[str] = None) -> Dict[str, Any]:
        import time
        start = time.time()
        try:
            client = AsyncOpenAI(base_url=proxy_url, api_key=api_key, timeout=30.0)
            models_resp = await client.models.list()
            models = [m.id for m in models_resp.data] if hasattr(models_resp, 'data') else []
            test_response = None
            if model:
                chat = await client.chat.completions.create(
                    model=model,
                    messages=[{'role': 'user', 'content': 'Say "hello" in one word.'}],
                    max_tokens=10,
                )
                test_response = chat.choices[0].message.content or ''
            return {
                'status': 'online',
                'latency_ms': int((time.time() - start) * 1000),
                'models': models,
                'test_response': test_response,
            }
        except Exception as e:
            return {
                'status': 'error',
                'latency_ms': int((time.time() - start) * 1000),
                'error': str(e),
            }
